services/supervisor/tests/supervisor/migrations/mod.rs

```diff
@@ -0,0 +1,879 @@
+
+    // Stop service instance before running the migration.
+    stop_service(&mut testkit, MigrationService::INSTANCE_ID);
+
+    // Request migration to 0.5.1.
+    // This migration will require three migration requests.
+    let deadline_height = DEADLINE_HEIGHT;
+    let request = MigrationRequest {
+        new_artifact: MigrationServiceV05_1.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    // After the first migration step, version should be "0.2".
+    wait_for_migration_success(
+        &mut testkit,
+        deadline_height,
+        request,
+        Version::new(0, 2, 0),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    migration_service::v02::verify_schema(prefixed);
+
+    // Request the same migration.
+    let deadline_height = Height(DEADLINE_HEIGHT.0 * 2);
+    let request = MigrationRequest {
+        new_artifact: MigrationServiceV05_1.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    // Now we should have version "0.5".
+    wait_for_migration_success(
+        &mut testkit,
+        deadline_height,
+        request,
+        Version::new(0, 5, 0),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    migration_service::v05::verify_schema(prefixed);
+
+    // Request the same migration for the third time.
+    let deadline_height = Height(DEADLINE_HEIGHT.0 * 3);
+    let request = MigrationRequest {
+        new_artifact: MigrationServiceV05_1.artifact_id(),
+        service: MigrationService::INSTANCE_NAME.into(),
+        deadline_height,
+    };
+
+    let api = testkit.api();
+    let tx_hash = request_migration(&api, request.clone());
+    let block = testkit.create_block();
+
+    block[tx_hash]
+        .status()
+        .expect("Transaction should be executed successfully");
+
+    // Now we finally should have version "0.5.1".
+    wait_for_migration_success(
+        &mut testkit,
+        deadline_height,
+        request,
+        Version::new(0, 5, 1),
+    );
+
+    let snapshot = testkit.snapshot();
+    let prefixed = Prefixed::new(MigrationService::INSTANCE_NAME, snapshot.as_ref());
+
+    // Data should not change.
+    migration_service::v05::verify_schema(prefixed);
+}
```

Comment:
Possible future work: It could make sense to test:

- Recovery after a failed migration (i.e., a migration fails, we restart it, it succeeds)
- Resuming a service after successful migration